### ISSUE TITLE
Remove onremove event.

### ISSUE THREAD

**Issue by jorgebucaran** [hyperapp/hyperapp#681]
There is no `onremove` equivalent in React/Preact/Vue that [I am aware of](https://github.com/jorgebucaran/ultradom/issues/68#issuecomment-382780383).

> **Note**: Hyperapp has the same problem as https://github.com/jorgebucaran/ultradom/issues/68. This can be and **will be fixed** by storing the element in the virtual node instead of using children position which is totally unreliable if you are adding / removing elements asynchronously.

**Comment by mindplay-dk** [comment]
Are you proposing to remove this for good, or just from the next release? (pending rewrite of patch?)

The feature isn't *useless* as it is - just unreliable.

It might make more sense to just log it as a bug, since it sounds like the patch rewrite will fix it? And the patch rewrite presumably is an internal change that doesn't need to be a breaking change - whereas removing the feature now is a breaking change.

**Comment by zaceno** [comment]
At least the way it works in petit-dom (the only vdom I’ve played much with besides Ultradom), as far as I can tell, the way you do any lifecycle events is to provide a component in the form of an object which defines the mount, unmount and patch methods.

When you do that, how the tree is handled from that point on is entirely up to you, including removing or deferring removal of the element.

As such, petit-dom’s “unmount” is closer to ultradom’s onremove than ondestroy.

I suspect it works similarly in React - or how else would exit transitions even be possible?

**Comment by jorgebucaran** [comment]
While investigating how to fix jorgebucaran/ultradom#68, I discovered two things: (1) how to fix the problem and (2) that there is no `onremove` equivalent in any of the mentioned frameworks.

The fix consists of storing the element in the virtual node as previously described, then retrieving it inside the async `done` handler. Even if there was no `onremove` I am still going to to store the element in the node at some point in the future when I have time to rewrite patch.

The reason I created this issue is because of (2) above. If others can get away without an `onremove` event, I'd like to explore how we could get rid of it too.

**Comment by infinnie** [comment]
But most other frameworks/libraries put lifecycle methods on *components* instead of DOM elements. And I think that is the difference.

**Comment by jorgebucaran** [comment]
@infinnie So, what's the difference?

**Comment by infinnie** [comment]
So if you want to put lifecycle methods on representations of DOM elements (which I actually think is better) and you want it to support custom element removal, you have to do something.

**Comment by infinnie** [comment]
And `onremove()` or the like is that “something”.

**Comment by localvoid** [comment]
Almost everyone who have worked on vdom algorithms had something similar to async `onremove` hook in their libraries, I even had a vdom implementation with an async create hook :) This hooks are useless, it is easy to solve all animation use cases without such hooks [1][2]

1. https://github.com/reactjs/react-transition-group
2. https://github.com/chenglou/react-motion

**Comment by jorgebucaran** [comment]
@localvoid I imagine all those frameworks you are referring to are component-based frameworks, which Hyperapp is not.

**Comment by infinnie** [comment]
While things like `componentWillUnmount()` are the equivalents in component-based frameworks. But they should be similar under the surface.

**Comment by jorgebucaran** [comment]
@infinnie Sorry, but I don't understand what you are saying.

> if you want to put lifecycle methods on representations of DOM element

We already have those, they are called [lifecycle events](https://github.com/hyperapp/hyperapp#lifecycle-events). 

> While things like componentWillUnmount() are the equivalents in component-based frameworks.

My understanding is that `componentWillUnmount` works not like Hyperapp's `onremove` but `ondestroy`, so they are not equivalent, that's [what I am talking about](https://github.com/hyperapp/hyperapp/issues/681#issue-316694083).

**Comment by infinnie** [comment]
@jorgebucaran thanks a lot.

**Comment by infinnie** [comment]
I didn’t negate that you already have lifecycle methods. It’s just that `ondestroy()` isn’t enough. Maybe we should seek for a userland solution instead, however.

**Comment by infinnie** [comment]
And @jorgebucaran you are right.

**Comment by zaceno** [comment]
@jorgebucaran If your plan is to store the element on the node, I suppose one option would be for someone who wants deferred removal to unassign it from the node. Then, when you go to remove the element, if it isn't still on the node, don't remove it.

 But you'd have to give access to the vnode in the ondestroy lifecycle handler... Maybe not a nice solution come to think of it.

**Comment by jorgebucaran** [comment]
@zaceno 

Clarifying: What I want to know is how to do without onremove by looking into how other libraries that have _no_ onremove equivalent do the same thing.

Storing the element in the node is how we will fix the current onremove (and allow improve the diff algo, so this is definitely happening at some point), so I could simply fix it and keep our onremove, but since react/preact don't have anything like it maybe we could also get away without

**Comment by zaceno** [comment]
I haven't read the source for e.g `react-transition-group`  or other similar solutions. But if React's VDOM does not support deferred removals, I can only assume their solution is based on transition-groups having local state (remembering the virtual node to animate out before removing it). 

Since local state is off the table, I suspect we need to keep `onremove`

**Comment by mindplay-dk** [comment]
What @zaceno said:

> if React's VDOM does not support deferred removals, I can only assume their solution is based on transition-groups having local state

That's my understanding - and, yes, from a developer's perspective, compared with the simplicity of the `onremove` event being a fundamental feature of element-level life-cycle events, the ultradom solution is frankly *much* more simple and elegant.

> Since local state is off the table, I suspect we need to keep `onremove`

Even if somebody were to build a local state layer using ultradom as the engine, it's still a *heck* of a lot more convenient than having to invent some new components/concepts and trying to shoehorn it in.

But IMO, the best, most fundamental argument in favor of `onremove` is simply this: *animation is an important part of the end-user experience.*

When used correctly, animation is not just fancy, fun eye-candy - for example, if you have a table with 20 rows, that split-second animation where a deleted row collapses and vanishes is there to give a visual indication ("experience") of *what* was removed.

How often do you see UI where everything janks and jumps when you add/remove something? It's confusing. You may not feel completely confident that you removed the correct item. "did I just remove the wrong one?" - it's unsettling, and that's why good designers use animation.

If you want to create a tool for quality UI, animation matters and `onremove` is a must-have.

**Comment by infinnie** [comment]
@mindplay-dk if only the `onremove`-out-of-order problem were solved.

**Comment by infinnie** [comment]
Ordering is a real problem the solution to which I look forward to.

**Comment by mindplay-dk** [comment]
> Ordering of elements being removed is a real problem the solution to which I look forward to

@infinnie note that `3.0.0` addresses the *general* reordering problem - a single re-render will have the elements ordered correctly now; problems only occur with deferred removals, so this is a step up at least 👍

**Comment by infinnie** [comment]
Just by the way: what would we expect to happen if two asynchronous element removals happen consecutively, so that when one element is being removed we start to remove the second one?

**Comment by infinnie** [comment]
When would the fix get released?

**Comment by jorgebucaran** [comment]
@infinnie After 2.0 along with https://github.com/hyperapp/hyperapp/issues/499 at some point :)

**Comment by infinnie** [comment]
Maybe I should be looking forward to it.

**Comment by jorgebucaran** [comment]
Maybe you should! 😄

**Comment by zaceno** [comment]
@frenzzy I don’t see how that could work in practice. Seems to me that means a lot of bookkeeping for the user, for every element they want transitioned. But maybe I don’t understand what you’re saying. By all means put together a demo of how that could work! I’d be interested in seeing it! (Honest, I’m not being sarcastic :) )

**Comment by frenzzy** [comment]
Pseudo code:
```js
// component-with-fade-out-effect.js
import animfx from 'hyperapp/animfx'

export const state = {
  anim: animfx.state.fadeOut,
  hidden: false
}

export const actions = {
  hide: () => ({ hidden: true })
}

export const view = ({ hidden, anim }) => hidden ? null : (
  <button style={anim.style} onclick={[animfx.actions.fadeOut, actions.hide]}>
    Click here to hide me with the fade out animation effect
  </button>
)
```

**Comment by frenzzy** [comment]
With lists it may works the similar way.
```js
// todo-with-animation.js
import animfx from 'hyperapp/animfx'

export const state = {
  anims: animfx.createState({
    type: 'list',
    effects: [animfx.fadeIn, animfx.fadeOut, animfx.move],
    uniqueKeyName: 'id'
  }),
  todos: []
}

export const actions = {
  add: (text) => (state) => [...state, { id: Math.random(), text, done: false }],
  del: (id) => (state) => state.filter(item => item.id !== id)
}

export const view = ({ todos, anims }) => hidden ? null : (
  <ul>
    {todos.map((todo) => (
      <li
        key={todo.id}
        style={anims[todo.id].style}
      >
        {todo.text}
        <button onclick={[anims[todo.id].fadeOut, actions.del]}>X</button>
      </li>
    )}
  </ul>
)
```

**Comment by jorgebucaran** [comment]
@frenzzy Why don't you build this? I am not sure I understand where you're going with this?

**Comment by infinnie** [comment]
And by the way: Storing DOM elements in the virtual node should not be relied on. Some virtual nodes are just strings, which are immutable and do not support custom properties.

**Comment by jorgebucaran** [comment]
@infinnie Yes, indeed. We're going to treat TextNodes as a virtual node too in the future.

**Comment by infinnie** [comment]
@jorgebucaran see also my edit.

**Comment by jorgebucaran** [comment]
Thanks. For example:

<pre>
{
  name: null,
  attributes: null,
  children: "Hello World",
  element: <samp>"Hello World"</samp>
}
</pre>
